### DISCUSSION TITLE
Revisit LuaJIT

### DISCUSSION THREAD

**Discussion by Fail-Safe** [sqm-autorate/sqm-autorate#76] [Operational Development]
Is there any significant reason to revisit an approach of switching back to LuaJIT since it's available for TurrisOS and OpenWrt?

**Comment by Lochnair** [comment]
I still think it's worth doing as the reduction in CPU usage is quite significant, so for those on platforms with not all that good processors, it'll likely make a difference.

**Comment by dlakelan** [comment]
What's needed to enable LuaJIT ?

**Reply by chas-iot** [reply]
There's a lack of easily findable resources on using LuaJIT
So far, the best I've found is http://wiki.luajit.org/Numerical-Computing-Performance-Guide
From other JIT technologies:
- use locals, not globals
- use explicit type conversions rather than implict (I'm not sure that this applies in lua)
- once a variable is used as one type, do not convert it to a different type. Don't do
- - `local x = 10`
- - `x = 'ten'`
- for frequently accessed functions create the upvalue `local max = math.max' at the module level rather than the function level

**Reply by Fail-Safe** [reply]
Just to make sure we're all on the same page, at one point @Lochnair and I were using JuaJIT for sqm-autorate. But we hit a segfault issue with LuaJIT + Lanes. Lanes has to be compiled against LuaJIT and @Lochnair got that working. However it's not currently out of the box functionality and was therefore a fairly significant hurdle for the average OpenWrt user.

**Reply by chas-iot** [reply]
[LuaJIT](https://github.com/LuaJIT/LuaJIT) seems to have restarted development under the original author

**Reply by trippleflux** [reply]
> Just to make sure we're all on the same page, at one point @Lochnair and I were using JuaJIT for sqm-autorate. But we hit a segfault issue with LuaJIT + Lanes. Lanes has to be compiled against LuaJIT and @Lochnair got that working. However it's not currently out of the box functionality and was therefore a fairly significant hurdle for the average OpenWrt user.

@Fail-Safe @Lochnair 
Is there any patches or mods for the lualanes OpenWRT package folder regarding lualanes compiled/linked with luajit ?

Currently my lualanes Makefile is this : 
```
define Package/lualanes
	SUBMENU:=Lua
	SECTION:=lang
	CATEGORY:=Languages
	TITLE:=LuaLanes
	URL:=http://lualanes.github.io/lanes/
	DEPENDS:=+luajit +liblua +libpthread
endef
```

**Reply by Lochnair** [reply]
@trippleflux I made it as a new package since I'm not sure if we can get one package to work for all Lua versions.
That plus an updated luajit package built from the latest Git at the time, which I can't recall the reason for now.

I put them over here: https://github.com/Lochnair/openwrt-feeds/tree/main/lang
So you should be able to add it to your feeds.conf to build those

**Reply by trippleflux** [reply]
@Lochnair 
Hmm I am getting segmentation fault, what is the correct luajit LUA_CPATH & LUA_PATH env vars?
In my .profile, I have : 
```
export LUA_CPATH='/root/.luarocks/lib/lua/5.1/?.so;/usr/lib/lua/5.1/?.so;./?.so;/usr/lib/lua/?.so;/usr/lib/lua/loadall.so;/usr/lib/lua/lanes/core.so;/usr/lib/libluajit-5.1.so'
export LUA_PATH='/root/.luarocks/share/lua/5.1/?.lua;/root/.luarocks/share/lua/5.1/?/init.lua;/usr/share/lua/5.1/?.lua;/usr/share/lua/5.1/?/init.lua;./?.lua;/usr/share/lua/?.lua;/usr/share/lua/?/init.lua;/usr/lib/lua/?.lua;/usr/lib/lua/?/init.lua'
```

If I have put /usr/lib/libluajit-5.1.so in the front of LUA_CPATH, I am getting : 
```
luajit: error loading module 'lanes.core' from file '/usr/lib/libluajit-5.1.so':
        /usr/lib/libluajit-5.1.so.2: undefined symbol: luaopen_lanes_core
stack traceback:
        [C]: at 0x7f0a904c0bd8
        [C]: in function 'require'
        /usr/lib/lua/lanes.lua:38: in main chunk
        [C]: in function 'require'
        sqm-autorate.lua:47: in main chunk
        [C]: at 0x00402981
```

my /usr/lib/lua/lanes/core.so is already containing  "luaopen_lanes_core" function.

[EDITED]
w00t, didn't know that there are different define Build/Compile in lualanes Makefile, recompiling snapshots, now.

**Reply by Lochnair** [reply]
> Hmm I am getting segmentation fault,

The segfault might be that you need the updated lualanes package, which I had to force install before installing the other packages to get it to use it:
```
./scripts/feeds update -a
./scripts/feeds install -p lochnair -f luajit
./scripts/feeds install -a
```

> what is the correct luajit LUA_CPATH & LUA_PATH env vars?

I tend to do this to set the environment when testing: `eval $(grep export /etc/init.d/sqm-autorate)` which grabs what we set in the service file, not the nicest way to do it, but it works


> w00t, didn't know that there are different define Build/Compile in lualanes Makefile, recompiling snapshots, now.
Not sure I'm following here, but I hope you figured it out :)

**Reply by trippleflux** [reply]
> > Hmm I am getting segmentation fault,
> 
> The segfault might be that you need the updated lualanes package, which I had to force install before installing the other packages to get it to use it:
> 
> ```
> ./scripts/feeds update -a
> ./scripts/feeds install -p lochnair -f luajit
> ./scripts/feeds install -a
> ```
> 
> > what is the correct luajit LUA_CPATH & LUA_PATH env vars?
> 
> I tend to do this to set the environment when testing: `eval $(grep export /etc/init.d/sqm-autorate)` which grabs what we set in the service file, not the nicest way to do it, but it works
> 
> > w00t, didn't know that there are different define Build/Compile in lualanes Makefile, recompiling snapshots, now.
> > Not sure I'm following here, but I hope you figured it out :)

@Lochnair 
Yes, Thanks now it's working fine!, I was using modified plain lualanes package as per-your shared OpenWRT packages, about luajit package, doesn't differ much from my mod.

Thanks also for the OpenWRT feeds tips.

About the /etc/config/sqm-autorotate details,
```
config network
        option upload_interface 'br-lan'
        option download_interface 'pppoe-wan'
        option upload_base_kbits '170000'
        option download_base_kbits '180000'
        option upload_min_percent '20'
        option download_min_percent '20'
```
I am connecting to my ISP through bridged pppoe client(pppoe-wan) in OpenWRT, and all internal LAN ports is bridged in "br-lan"

in sqm-autorotate start-up, I am getting : 
```
[ head -n 10 /tmp/sqm-autorate.log ]
RTNETLINK answers: Invalid argument
RTNETLINK answers: No such file or directory
RTNETLINK answers: Invalid argument
RTNETLINK answers: No such file or director
```
Does my "upload_interface" & "download_interface" are already proper setup?

**Comment by chas-iot** [comment]
@trippleflux what is the output of `tc qdisc | grep cake` on your router?

**Reply by trippleflux** [reply]
> @trippleflux what is the output of `tc qdisc | grep cake` on your router?

@chas-iot
Hmm empty, returning no reply/output.

**Reply by trippleflux** [reply]
> > @trippleflux what is the output of `tc qdisc | grep cake` on your router?
> 
> @chas-iot Hmm empty, returning no reply/output.

If I set "Queue setup script" in the OpenWRT webui as "layer_cake.qos" then replied as the following : 
```
root@OpenWRT:/usr/lib/sqm-autorate# tc qdisc | grep cake
qdisc cake 800a: dev pppoe-wan root refcnt 2 bandwidth 180Mbit diffserv3 triple-isolate nonat nowash no-ack-filter split-gso rtt 100ms raw overhead 0
qdisc cake 800b: dev ifb4pppoe-wan root refcnt 2 bandwidth 170Mbit besteffort triple-isolate nonat wash no-ack-filter split-gso rtt 100ms raw overhead 0
```

before I have set it as "simplest_tbf.qos"

**Reply by trippleflux** [reply]
> > > @trippleflux what is the output of `tc qdisc | grep cake` on your router?
> > 
> > 
> > @chas-iot Hmm empty, returning no reply/output.
> 
> If I set "Queue setup script" in the OpenWRT webui as "layer_cake.qos" then replied as the following :
> 
> ```
> root@OpenWRT:/usr/lib/sqm-autorate# tc qdisc | grep cake
> qdisc cake 800a: dev pppoe-wan root refcnt 2 bandwidth 180Mbit diffserv3 triple-isolate nonat nowash no-ack-filter split-gso rtt 100ms raw overhead 0
> qdisc cake 800b: dev ifb4pppoe-wan root refcnt 2 bandwidth 170Mbit besteffort triple-isolate nonat wash no-ack-filter split-gso rtt 100ms raw overhead 0
> ```
> 
> before I have set it as "simplest_tbf.qos"

my "ip a" output : 
```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq master br-lan state UP group default qlen 1000
    link/ether 40:62:31:0b:fc:5b brd ff:ff:ff:ff:ff:ff
3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq master br-lan state UP group default qlen 1000
    link/ether 40:62:31:0b:fc:5c brd ff:ff:ff:ff:ff:ff
4: eth2: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq master br-lan state UP group default qlen 1000
    link/ether 40:62:31:0b:fc:5d brd ff:ff:ff:ff:ff:ff
5: eth3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq master br-lan state UP group default qlen 1000
    link/ether 40:62:31:0b:fc:5e brd ff:ff:ff:ff:ff:ff
6: eth4: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq master br-lan state UP group default qlen 1000
    link/ether 40:62:31:0b:fc:5f brd ff:ff:ff:ff:ff:ff
7: eth5: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 40:62:31:0b:fc:60 brd ff:ff:ff:ff:ff:ff
8: ip6tnl0@NONE: <NOARP> mtu 1452 qdisc noop state DOWN group default qlen 1000
    link/tunnel6 :: brd :: permaddr 6aa0:0:c7b6::
9: sit0@NONE: <NOARP> mtu 1480 qdisc noop state DOWN group default qlen 1000
    link/sit 0.0.0.0 brd 0.0.0.0
10: gre0@NONE: <NOARP> mtu 1476 qdisc noop state DOWN group default qlen 1000
    link/gre 0.0.0.0 brd 0.0.0.0
11: gretap0@NONE: <BROADCAST,MULTICAST> mtu 1462 qdisc noop state DOWN group default qlen 1000
    link/ether 00:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff
12: erspan0@NONE: <BROADCAST,MULTICAST> mtu 1450 qdisc noop state DOWN group default qlen 1000
    link/ether 00:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff
13: ip6gre0@NONE: <NOARP> mtu 1448 qdisc noop state DOWN group default qlen 1000
    link/gre6 :: brd :: permaddr f61c:35aa:1fd3::
14: teql0: <NOARP> mtu 1500 qdisc noop state DOWN group default qlen 100
    link/void
15: br-lan: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether 40:62:31:0b:fc:5b brd ff:ff:ff:ff:ff:ff
    inet 10.10.8.1/21 brd 10.10.15.255 scope global br-lan
       valid_lft forever preferred_lft forever
    inet6 fd4c:cb3f:68b0::1/60 scope global noprefixroute
       valid_lft forever preferred_lft forever
    inet6 fe80::4262:31ff:fe0b:fc5b/64 scope link
       valid_lft forever preferred_lft forever
21: pppoe-wan: <POINTOPOINT,MULTICAST,NOARP,UP,LOWER_UP> mtu 1492 qdisc cake state UNKNOWN group default qlen 3
    link/ppp
    inet 123.123.123.123 peer 123.123.123.123/32 scope global pppoe-wan
       valid_lft forever preferred_lft forever
    inet6 ipv6/128 scope link
       valid_lft forever preferred_lft forever
28: ifb4pppoe-wan: <BROADCAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc cake state UNKNOWN group default qlen 32
    link/ether 1a:14:bb:c8:63:3d brd ff:ff:ff:ff:ff:ff
    inet6 fe80::1814:bbff:fec8:633d/64 scope link
       valid_lft forever preferred_lft forever
```

**Reply by chas-iot** [reply]
> If I set "Queue setup script" in the OpenWRT webui as "layer_cake.qos" then replied as the following :
> ```
> root@OpenWRT:/usr/lib/sqm-autorate# tc qdisc | grep cake
> qdisc cake 800a: dev pppoe-wan root refcnt 2 bandwidth 180Mbit diffserv3 triple-isolate nonat nowash no-ack-filter split-gso rtt 100ms raw overhead 0
> qdisc cake 800b: dev ifb4pppoe-wan root refcnt 2 bandwidth 170Mbit besteffort triple-isolate nonat wash no-ack-filter split-gso rtt 100ms raw overhead 0
> ```
> before I have set it as "simplest_tbf.qos"

With cake setup like this, `sqm-autorate should have
```
option upload_interface 'pppoe-wan'
option download_interface 'ifb4pppoe-wan'
```
It just won't work unless cake is setup.

**Reply by trippleflux** [reply]
Thanks, now I understand, this means cake setup for "layer_cake.qos" right? or simplest?.
> option upload_interface 'pppoe-wan'
> option download_interface 'ifb4pppoe-wan'

**Reply by Lochnair** [reply]
`layer_cake` or `piece_of_cake` is the the ones you want yeah :)

**Reply by trippleflux** [reply]
Thanks, now fully getting it , I think this should be documented if not already.

**Reply by Lochnair** [reply]
I don't think it is, probably because we've assumed that anyone trying this already have a CAKE setup working,
but yeah it'd probably be a good idea to make sure this is clear in the README. Which by the way is getting a good rework over here #113, so if you have suggestions please add them there

**Reply by trippleflux** [reply]
@Lochnair 

Done, but I didn't open a pull request.

Hopefully it will be helpful.

**Comment by Lochnair** [comment]
Decided to take a little look at this again. One of the LuaJIT problems is caused by the Lua 5.1 openwrt package having a patch adding the lua_isinteger function that wasn't actually introduced until later. When Lanes is compiled it sees that the function is available and so expects that also when run with LuaJIT.

I wrote a little patch for LuaJIT to introduce that function there as well since it's a simple one, but that wasn't enough.
Next problem is the 'bad light userdata pointer' issue. Still no clue what that's actually caused by. IIRC what solved it previously was building from LuaJIT git, but that's not tagged and I'm not going to bother digging through the commits to find what fixed it.

So I decided to give https://github.com/openresty/luajit2 a try, they at least tag new versions from time to time.
Changed the package to build from the latest tag there, and it runs normally for me on x86

**Reply by Fail-Safe** [reply]
@Lochnair I've been looking into this again. I tried to compile https://github.com/LuaJIT/LuaJIT/commit/c525bcb9024510cad9e170e12b6209aedb330f83 against lualanes 3.16.2, but I keep hitting this error:

```
cc1: note: someone does not honour COPTS correctly, passed 0 times
In file included from deep.h:11,
                 from tools.h:6,
                 from tools.c:43:
tools.c:1511:34: error: 'CLONABLES_CACHE_KEY' defined but not used [-Werror=unused-const-variable=]
 1511 | static DECLARE_CONST_UNIQUE_KEY( CLONABLES_CACHE_KEY, 0xD04EE018B3DEE8F5);
      |                                  ^~~~~~~~~~~~~~~~~~~
uniquekey.h:20:62: note: in definition of macro 'DECLARE_CONST_UNIQUE_KEY'
   20 | #define DECLARE_CONST_UNIQUE_KEY( name_, p_) UniqueKey const name_ = { MAKE_UNIQUE_KEY( p_)}
      |                                                              ^~~~~
cc1: all warnings being treated as errors
make[5]: *** [<builtin>: tools.o] Error 1
make[5]: Leaving directory '/home/user/openwrt/build_dir/target-x86_64_musl/lanes-3.16.2/src'
make[4]: *** [Makefile:46: src/lanes/core.so] Error 2
```

I'm afraid I'm too rusty to make enough sense of it to get past that.

Did luajit2 require anything special to work with lualanes?

**Reply by Fail-Safe** [reply]
I finally got lanes to compile with luajit. It's pretty incredible how much lower CPU utilization is with luajit vs lua.

**Reply by Lochnair** [reply]
@Fail-Safe Good to hear :)
What ended up being the solution? And yeah the difference is really significant, so it makes Lua much more viable on lower end hardware

**Reply by Fail-Safe** [reply]
It may have just been a "me" problem in that I was trying to get my head wrapped back around all of the lua/lanes/luajit nuances. Regardless, the changes I made that are working for me right now are summed up in this PR: https://github.com/Lochnair/openwrt-feeds/pull/1

**Reply by Fail-Safe** [reply]
FWIW, some general numbers I am seeing are that lua hit a 1.1% cumulative CPU usage figure after about 20 minutes of operation:
```
root@OpenWrt:~# ps -C lua -o %cpu,%mem,cmd
%CPU %MEM CMD
 1.1  0.0 /usr/bin/lua /usr/lib/sqm-autorate/sqm-autorate.lua
```

Whereas luajit is sitting at about 0.8% cumulative CPU usage after about 18 hours of steady operation:
```
root@OpenWrt:~# ps -C luajit -o %cpu,%mem,cmd
%CPU %MEM CMD
 0.8  0.0 /usr/bin/luajit /usr/lib/sqm-autorate/sqm-autorate.lua
```

**Reply by Lochnair** [reply]
Gotcha, I've gone ahead and merged that :)

**Reply by Lochnair** [reply]
You've got a pretty beefy box as a router don't you? So would probably show even better on something more low powered

**Reply by Fail-Safe** [reply]
Yeah, I'm on a 4core Intel N100 (no HT) these days and it holds its own 😎

Would love to see if someone on an older CPU could do some benchmarks as well. Kind of a ghost town around here at the moment, though.

I have been really impressed with the way @chas-iot's delay-histogram plugin is working on my FTTH connection. Wish I had tried it sooner.

**Reply by Lochnair** [reply]
Aha we both went the same route with the N100 😄 
Do you have one of those Chinese brands as well or?

It has been yeah, and now I too have moved to FTTH, and (un)luckily mine is stable enough that autorate isn't really needed.
But of course, I don't mind testing that it doesn't make things worse on my connection if desired :)

**Reply by Fail-Safe** [reply]
Indeed I do--I picked up a Beelink EQ12. It "just works" and does so pretty darn power efficiently

I really don't "need" any sort of autorate functionality. However, being on GPON means I can at times of the day still hit a bit of contention when my neighbors are also pushing hard on upload/download. For those times of day, having something like sqm-autorate able to make some minor adjustments to keep latency as low as possible is my ultimate goal.

Beyond my selfish reasons for an autorate... I generally was always impressed with how well this lua implementation worked and would love for "the world" to still have it available for those who find it useful. :)

**Reply by Lochnair** [reply]
Nifty little devices those, mine is from WooYi, though I guess that's probably not the OEM. And they're absolutely very power efficient devices, and surprisingly powerful at that.

I _think_ I'm on GPON too, but either way haven't really noticed any slowdowns that's affected me, so I don't care too much 😄 
And either way I'll be moving to another vendor soon enough, that I believe doesn't use GPON. Sadly they use CGNAT, but they do offer "static" public IP's for a little extra money, so I can live with that.

But yes even though I don't need it anymore, I would like for it to actually work without people having to patch lualanes in openwrt to make it work :)
